torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -347,6 +356,300 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  */
+  int64_t ndim = PyArray_NDIM(array);
+  npy_intp* array_shape = PyArray_DIMS(array);
+  for (int64_t i = 0; i < ndim; ++i) {
+    shape.emplace_back(array_shape[i]);
+  }
+}
+
+bool _extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, std::vector<int64_t> & ndarray_shape, int64_t shape_idx, std::vector<PyArrayObject*> & array_ptr_storage) {
+  /*
+  This is a helper function for extracting all the numpy arrays embedded within PySequences (recursively) representd by obj. This is done recursively in a 
+  depth first search style. It puts the numpy array pointers in the array_ptr_storage for later use. In case there is a non numpy array type within obj, 
+  this function immediately returns false. If only all the items inside obj are just numpy arrays, it returns true. It throws a ValueError in case of a dimension
+  mismatch. This function is used in extract_ndarrays() function below.
+  */
+  if (PyArray_Check(obj)) {
+    if (PyArray_NDIM((PyArrayObject*) obj) == 0) {
+      return false;
+    }
+
+    if (ndarray_shape.size() == 0) {
+      ndarray_shape.emplace_back(shape_idx);
+      _populate_shape_vector(ndarray_shape, (PyArrayObject*) obj);
+    }
+    else if ((ndarray_shape[0] != shape_idx) || !_shape_match_ndarray(ndarray_shape, (PyArrayObject*) obj, 1)) {
+      auto true_shape_str = get_buffer_str<npy_intp> (static_cast<void*>(PyArray_DIMS((PyArrayObject*) obj)), 0, PyArray_NDIM((PyArrayObject*) obj));
+      auto expected_shape_str = get_vector_str(ndarray_shape, 1);
+      std::ostringstream err;
+      err << "expected numpy array of shape " << expected_shape_str <<" at dim " << ndarray_shape[0] << " (got " << true_shape_str << " at dim " << shape_idx << ")";
+      throw ValueError(err.str());  
+    }
+    array_ptr_storage.push_back((PyArrayObject*) obj);
+    return true;
+  } else if (PySequence_Check(obj)) {
+    auto seq = THPObjectPtr(PySequence_Fast(obj, "not a sequence"));
+    if (!seq) {
+      throw python_error();
+    }
+    Py_ssize_t seq_len = PySequence_Fast_GET_SIZE(seq.get());
+    if (seq_len != shape[shape_idx]) {
+      throw ValueError("expected sequence of length %lld at dim %lld (got %lld)",
+        (long long)shape[shape_idx], (long long)shape_idx, (long long)seq_len);
+    }
+    PyObject** items = PySequence_Fast_ITEMS(seq.get());
+    for (Py_ssize_t i = 0; i < seq_len; ++i) {
+      if (!_extract_ndarrays(items[i], shape, ndarray_shape, shape_idx+1, array_ptr_storage)) {
+        return false;
+      }
+    }
+  } else {
+    return false;
+  }
+  return true;
+}
+
+
+bool extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, std::vector<PyArrayObject*> & array_ptr_storage) {
+  /*
+  A function for extracting all numpy arrays embedded (recursively) within PySequences in obj. Failure to do so will return
+  false, else true. The extracted numpy array pointers are stored in array_ptr_storage. It uses _extract_ndarrays() function 
```

Comment:
Why would this return false?

Reply:
The only time it will return true is when all the items are non-scalar numpy arrays having the same shapes and also all of them need to be the leaf nodes at the same height.  So, ```[[a,b], c]``` will return false as ```a``` and ```c``` are at different heights.

Reply:
Thank you, that's helpful. Please add that to the comment, too.